httpcore5/src/main/java/org/apache/hc/core5/http/protocol/RequestTE.java

```diff
@@ -0,0 +1,190 @@
+     * @throws IOException   in case of an I/O error
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context)
+            throws HttpException, IOException {
+        Args.notNull(request, "HTTP request");
+
+        // Fetch the TE header
+        final Header teHeader = request.getFirstHeader(HttpHeaders.TE);
+
+        if (teHeader == null) {
+            return;  // No further validation needed
+        }
+
+        final String teValue = teHeader.getValue();
+        validateTEField(teValue);
+
+        validateConnectionHeader(request);
+    }
+
+    /**
+     * Validates the {@code TE} header values for compliance with HTTP/1.1.
+     * <p>
+     * Specifically, this method ensures that:
+     * <ul>
+     * <li>The {@code TE} header does not contain the {@code chunked} transfer coding.</li>
+     * <li>The {@code trailers} directive is allowed and treated as valid.</li>
+     * </ul>
+     *
+     * @param teValue the value of the {@code TE} header
+     * @throws HttpException if the {@code TE} header contains invalid values
+     */
+    private void validateTEField(final String teValue) throws HttpException {
+        final Tokenizer.Cursor cursor = new Tokenizer.Cursor(0, teValue.length());
+
+        while (!cursor.atEnd()) {
+            Tokenizer.INSTANCE.skipWhiteSpace(teValue, cursor);
+
+            final String member = Tokenizer.INSTANCE.parseToken(teValue, cursor, DELIMITER);
+
+            if (member.isEmpty()) {
+                if (!cursor.atEnd()) {
+                    Tokenizer.INSTANCE.skipWhiteSpace(teValue, cursor);
+                    cursor.updatePos(cursor.getPos() + 1);
+                }
+                continue;
+            }
+
+            if ("trailers".equalsIgnoreCase(member)) {
+                continue;
+            }
+
+            if (HeaderElements.CHUNKED_ENCODING.equalsIgnoreCase(member)) {
+                throw new ProtocolException("'chunked' transfer coding must not be listed in the TE header for HTTP/1.1.");
+            }
+
+            if (!cursor.atEnd()) {
+                Tokenizer.INSTANCE.skipWhiteSpace(teValue, cursor);
+                cursor.updatePos(cursor.getPos() + 1);
+            }
+        }
+    }
+
+    /**
+     * Validates the presence of the {@code Connection: TE} header when the {@code TE} header is present.
+     * <p>
+     * If the {@code TE} header is used, the HTTP/1.1 protocol requires that the {@code Connection} header includes the {@code TE} directive to prevent forwarding by intermediaries.
+     *
+     * @param request the HTTP request to validate
+     * @throws HttpException if the {@code Connection: TE} header is missing
+     */
+    private void validateConnectionHeader(final HttpRequest request) throws HttpException {
+        final Header connectionHeader = request.getFirstHeader(HttpHeaders.CONNECTION);
```

Comment:
@arturobernalg Please also note it is legal for a message to have multiple `Connection` headers. Just looking at the first one is not good enough. I think the same applies to `TE`